### DISCUSSION TITLE
How to build and run this repository?

### DISCUSSION THREAD

**Discussion by Kasper-This** [ros-industrial-consortium/scan_n_plan_workshop#112] [General]
Im sorry if this may be a stupid Question, but I'm pretty much a beginner in ROS so I'm easily confused.
What ROS version does this run in? Does this run through foxy?
Also, how do i install this package and run one of the examples? Do i have to merge the packages into one singe package and then run the colcon build?

**Comment by marrts** [comment]
The best way to build and run this is by following the [build instructions in the README](https://github.com/ros-industrial-consortium/scan_n_plan_workshop?tab=readme-ov-file#build-setup). This works in both ROS 2 Foxy and Humble.

To run the system you need to also include one of the SNP implementations in your workspace. You can see how to do that in this other repository [here](https://github.com/ros-industrial-consortium/snp_automate_2023?tab=readme-ov-file#local-build). You'll clone that other repo into your same workspace, resulting in two packages in your workspace.

So you'll first clone both this repo and the other one I linked and your workspace directories should be structured like this:
```
your_colcon_ws
└── src
    ├── scan_n_plan_workshop
    ├── snp_automate_2023
```

Then you'll run all the vcs import commands (if you have any issues with this look [here](http://wiki.ros.org/vcstool)) and your workspace should look something like this (Note: I may have some extra or missing packages):
```
your_colcon_ws
└── src
    ├── BehaviorTree.CPP
    ├── BehaviorTree.ROS2
    ├── boost_plugin_loader
    ├── descartes_light
    ├── ifopt
    ├── industrial_core_ros2_msgs_only
    ├── industrial_reconstruction
    ├── motoman
    ├── motoros2_interfaces
    ├── noether
    ├── opw_kinematics
    ├── realsense-ros
    ├── robot_cal_tools
    ├── ros_industrial_cmake_boilerplate
    ├── rviz_polygon_selection_tool
    ├── scan_n_plan_workshop
    ├── snp_automate_2023
    ├── taskflow
    ├── tesseract
    ├── tesseract_planning
    ├── tesseract_qt
    ├── tesseract_ros2
    ├── trajectory_preview
    └── trajopt
```
And then at your base level workspace directory (`your_colcon_ws`) you'll do:

Make sure you have all your ROS dependencies installed (if you have any issues look [here](http://wiki.ros.org/rosdep))
```
rosdep install --from-paths src -iry --skip-keys libvtk
```
 Build
```
colcon build --cmake-args -DTESSERACT_BUILD_FCL=OFF -DBUILD_RENDERING=OFF
```

If you're familiar with Docker, you can look at how [we are doing things in the Dockerfile](https://github.com/ros-industrial-consortium/scan_n_plan_workshop/blob/master/docker/Dockerfile) for building this repo as reference. 

You might also need to run
```
python3 -m pip install open3d pyquaternion numpy --upgrade
```

**Reply by Kasper-This** [reply]
WOW! Thank you very much for this detailed answer!
This helps a lot. I'll keep you posted when i succeed

**Reply by marrts** [reply]
No problem. I know how overwhelming it can be when you're first getting started with all of this stuff.

**Reply by Kasper-This** [reply]
So im trying to install the dependencies atm. And but the install in the README [here](https://github.com/ros-industrial-consortium/scan_n_plan_workshop) in point 2) says to install the packages into the workspace folder, not the src folder one level deeper, right?

 cd <snpd_workspace>
 vcs import < src/scan_n_plan_workshop/dependencies_tesseract.repos
 vcs import < src/scan_n_plan_workshop/dependencies.repos
 rosdep install --from-paths src --ignore-src -r -y`

I think that was one potential error source for why i could not run the rosdep install command.

**Reply by marip8** [reply]
Yes, I think that's a typo. I think it should be:

```diff
cd <snpd_workspace>
-vcs import < src/scan_n_plan_workshop/dependencies_tesseract.repos
+vcs import src < src/scan_n_plan_workshop/dependencies_tesseract.repos
-vcs import < src/scan_n_plan_workshop/dependencies.repos
+vcs import src < src/scan_n_plan_workshop/dependencies.repos
rosdep install --from-paths src --ignore-src -r -y
```

**Comment by Kasper-This** [comment]
Another question. Is this the right thread to ask? I'm not even sure if this is an issue specific to your library. 

so it seems i have still issues trying to install and build all the dependencies. 
Specific packages do not detect others `XYZ-Package: Cannot locate rosdep definition for [ABC_Package]`
This seems to be solved by individually building the necessary package before installing the former. Maybe there is an easier solution than mine, but I'm still pretty new to this. 

Another problem seemed to be contradictory dependencies? So the package noether requires libvtk7. But as soon as Ubuntu automatically wants to install the library through `sudo -H apt-get install -y libvtk7-dev` I get the following error: `python3-vtk9 : Breaks: python3-vtk7 but 7.1.1+dfsg2-10.1build1 is to be installed`

I tried downgrading python3-vtk9 to python-vtk7. But it seems to automatically upgrade again when i try to build the workspace. 
Have you encountered a similar problem before?

**Reply by Tobimane** [reply]
Hi, i am facing the exact same problem. Were you able to solve that?

Cheers,
T

**Comment by hariharan382** [comment]
Hi, I have set everything ready, I could build every package, but how do I get started with this pkg, how and where is the mesh file that is scanned, I tried the snp_blending launch "start.launch.xml", which is not showing up the simulated-real sense camera topics, and how should I use the snp_application app. Is there any documentation for this package, I could only find the node computation graph image(PNG) in docs, Thanks

**Reply by marip8** [reply]
This repository no longer contains the actual implementations for specific hardware setups. We have created a few of these repos, and they are linked on the [README](https://github.com/ros-industrial-consortium/scan_n_plan_workshop?tab=readme-ov-file#application-specific-implementations). They have some nominal instructions for downloading/building and running them in simulation and on hardware (if you happened to have all of the same hardware). `snp_automate_2023` is probably the latest and most up-to-date of these implementation repositories. I would recommend starting with that one and running it in simulation to understand how it works. [This video](https://www.youtube.com/watch?v=GgTxvIaekjE) is a bit dated now, but it gives a run through of the architecture of SNP and how to run one of the implementations (they all function essentially the same way)
